connman/src/storage.c

```diff
@@ -794,11 +796,58 @@ static void keyfile_cleanup(void)
 	keyfile_hash = NULL;
 }
 
+static int validate_path(const char *pathname)
+{
+	char *expanded_path;
+	int err = 0;
+
+	/* realpath() expands all symbolic links and references. */
+	expanded_path = realpath(pathname, NULL);
+	if (!expanded_path) {
+		switch (errno) {
+		case EACCES:
+		case EINVAL: /* pathname was NULL */
+		case EIO:
+		case ELOOP:
+		case ENAMETOOLONG:
+		case ENOMEM:
+		case ENOTDIR:
+			DBG("realpath() error %d/%s", errno, strerror(errno));
+			return -errno;
+		/* When saving the file does not exist */
+		case ENOENT:
```

Comment:
Note: ENOENT will be returned for symlink-to-missing. Not sure if that is an issue, but I think you could at least store data in another user's data dir.

Reply:
[man page](https://www.man7.org/linux/man-pages/man3/realpath.3.html) only states that `ENOENT The named file does not exist.` Are you referring to the:
```
GNU extensions
       If the call fails with either EACCES or ENOENT and resolved_path
       is not NULL, then the prefix of path that is not readable or does
       not exist is returned in resolved_path.
```
In which case `errno` should be checked also when the `expanded_path` is set.

Reply:
The man page didn't give enough information, so I created a trivial test case to see how it works :)

The GNU extension can indeed be used to check for the symlink-to-missing, but you'll need to use the buffer argument to get the info.

So with a symlink-to-missing, you'll get: rv: NULL, buffer will be filled with the name of the linked file, errno ENOENT

Reply:
Oh joy, world shall burn when you cannot trust man pages ;)

Reply:
Actually it seems that `g_file_set_contents()` does fail if there is a symlink to missing location for a service about to be saved in `storage_save()`. But anyways, made a reformatting to be more clear in call cases (hopefully).